pkg/gauges/contracts/gauges/StakelessGauge.sol

```diff
@@ -0,0 +1,143 @@
+        uint256 currentPeriod = _currentPeriod();
+
+        if (lastPeriod < currentPeriod) {
+            _gaugeController.checkpoint_gauge(address(this));
+
+            uint256 rate = _rate;
+            uint256 newEmissions = 0;
+            lastPeriod += 1;
+            uint256 nextEpochTime = _startEpochTime + _RATE_REDUCTION_TIME;
+            for (uint256 i = lastPeriod; i < lastPeriod + 255; ++i) {
+                if (i > currentPeriod) break;
+
+                uint256 periodTime = i * 1 weeks;
+                uint256 periodEmission = 0;
+                uint256 gaugeWeight = _gaugeController.gauge_relative_weight(address(this), periodTime);
+
+                if (nextEpochTime >= periodTime && nextEpochTime < periodTime + 1 weeks) {
+                    // If the period crosses an epoch, we calculate a reduction in the rate
+                    // using the same formula as used in `BalancerTokenAdmin`. We perform the calculation
+                    // locally instead of calling to `BalancerTokenAdmin.rate()` because we are generating
+                    // the emissions for the upcoming week, so there is a possibility the new
+                    // rate has not yet been applied.
+
+                    // Calculate emission up until the epoch change
+                    periodEmission = (gaugeWeight * rate * (nextEpochTime - periodTime)) / 10**18;
+                    // Action the decrease in rate
+                    rate = (rate * _RATE_DENOMINATOR) / _RATE_REDUCTION_COEFFICIENT;
+                    // Calculate emission from epoch change to end of period
+                    periodEmission += (gaugeWeight * rate * (periodTime + 1 weeks - nextEpochTime)) / 10**18;
+
+                    _rate = rate;
+                    _startEpochTime = nextEpochTime;
+                    nextEpochTime += _RATE_REDUCTION_TIME;
+                } else {
+                    periodEmission = (gaugeWeight * rate * 1 weeks) / 10**18;
+                }
+
+                emit PeriodEmission(periodTime, periodEmission);
+                newEmissions += periodEmission;
+            }
+
+            _period = currentPeriod;
+            _emissions += newEmissions;
+
+            if (newEmissions > 0 && !isKilled) {
+                _minter.mint(address(this));
+                _postMintAction(newEmissions);
+            }
+        }
+
+        return true;
+    }
+
+    function _currentPeriod() internal view returns (uint256) {
+        return (block.timestamp / 1 weeks) - 1;
+    }
+
+    function _postMintAction(uint256 mintAmount) internal virtual;
+
+    function user_checkpoint(address) external pure override returns (bool) {
+        return true;
+    }
+
+    function integrate_fraction(address user) external view override returns (uint256) {
+        require(user == address(this), "Gauge can only mint for itself");
+        return _emissions;
+    }
+
+    function set_killed(bool _isKilled) external {
```

Comment:
We're missing any authentication here. I'm slightly torn between using the AuthorizerAdaptor (for consistency with other gauges) vs just using the Authorizer natively although I'm leaning towards the former, thoughts?